N°3448 - Framework field size check not correctly implemented for multibytes languages/strings

accognet opened the pull request:
internal

accognet committed:
Replace strlen with mb_strlen. Better but without big impact (only the logs and the display are affected by this change)

accognet committed:
N°3448 - Framework field size check not correctly implemented for multibytes languages/strings

accognet committed:
Replace strlen with mb_strlen. Better but without big impact (only eventIssues are affected by this change)

Hipska commented on synchro/synchrodatasource.class.inc.php:
```diff
@@ -1938,6 +1938,7 @@ protected function TraceToText()
 		{
 			$sTrace = implode("\n", $this->m_aTraces);
 		}
+		//YOYO or not ?
```
YOYO? 🤓

accognet replied:
Yes, modify or not ? I'm not familiar with this code and I don't want to forget to discuss it 😁

Hipska replied:
I get the meaning, but what does the YOYO abbreviation stand for?

piRGoif replied:
You may instead use a classic TODO or FIXME, they are recognized by default by most IDE including PHPStorm.
Maybe something like : `//TODO PR #528 to discuss : should we modify or not ?`

Hipska replied:
Is it a French abbreviation perhaps?

accognet replied:
No, I'm the only one using that word. It's easier to find in the code than a meaningful word :-).

Hipska commented on webservices/webservices.class.inc.php:
```diff
@@ -298,8 +298,7 @@ protected function TrimAndSetValue($oLog, $sAttCode, $sValue)
-			if ($iMaxSize && (strlen($sValue) > $iMaxSize))
-			{
+			if ($iMaxSize && (mb_strlen($sValue) > $iMaxSize)) {
 				$sValue = substr($sValue, 0, $iMaxSize);
```
```suggestion
				$sValue = mb_substr($sValue, 0, $iMaxSize);
```

Hipska commented on synchro/synchrodatasource.class.inc.php:
```diff
@@ -2145,8 +2146,7 @@ protected function RecordWarnings()
-		if (strlen($sWarningMessage) > $MAX_WARNING_LENGTH)
-		{
+		if (mb_strlen($sWarningMessage) > $MAX_WARNING_LENGTH) {
 			$sWarningMessage = substr($sWarningMessage, 0, $MAX_WARNING_LENGTH - 3).'...';
```
```suggestion
			$sWarningMessage = mb_substr($sWarningMessage, 0, $MAX_WARNING_LENGTH - 3).'...';
```

Hipska commented on synchro/synchrodatasource.class.inc.php:
```diff
@@ -2187,16 +2187,12 @@ protected function DBDeleteSingleObject()
-		if (strlen($sText) > 255)
-		{
+		if (mb_strlen($sText) > 255) {
 			$sText = substr($sText, 0, 200).'...('.strlen($sText).' chars)...';
```
```suggestion
			$sText = mb_substr($sText, 0, 200).'...('.strlen($sText).' chars)...';
```

Hipska commented on core/event.class.inc.php:
```diff
@@ -242,44 +242,33 @@ protected function OnInsert()
-			$this->Set('issue', substr($this->Get('issue'), 0, 200)." -truncated ($sLength chars)");
+		$sLength = mb_strlen($this->Get('issue'));
+		if ($sLength > 255) {
+			$this->Set('issue', substr($this->Get('issue'), 0, 210)." -truncated ($sLength chars)");
```
```suggestion
			$this->Set('issue', mb_substr($this->Get('issue'), 0, 210)." -truncated ($sLength chars)");
```

Hipska commented on core/event.class.inc.php:
```diff
@@ -242,44 +242,33 @@ protected function OnInsert()
-			$this->Set('impact', substr($this->Get('impact'), 0, 200)." -truncated ($sLength chars)");
+		$sLength = mb_strlen($this->Get('impact'));
+		if ($sLength > 255) {
+			$this->Set('impact', substr($this->Get('impact'), 0, 210)." -truncated ($sLength chars)");
```
```suggestion
			$this->Set('impact', mb_substr($this->Get('impact'), 0, 210)." -truncated ($sLength chars)");
```

Hipska commented on core/event.class.inc.php:
```diff
@@ -242,44 +242,33 @@ protected function OnInsert()
-			$this->Set('page', substr($this->Get('page'), 0, 200)." -truncated ($sLength chars)");
+		$sLength = mb_strlen($this->Get('page'));
+		if ($sLength > 255) {
+			$this->Set('page', substr($this->Get('page'), 0, 210)." -truncated ($sLength chars)");
```
```suggestion
			$this->Set('page', mb_substr($this->Get('page'), 0, 210)." -truncated ($sLength chars)");
```

Hipska commented on core/dbobject.class.php:
```diff
@@ -746,8 +746,7 @@ public function SetTrim($sAttCode, $sValue)
-		if ($iMaxSize && (strlen($sValue) > $iMaxSize))
-		{
+		if ($iMaxSize && (mb_strlen($sValue) > $iMaxSize)) {
 			$sValue = substr($sValue, 0, $iMaxSize);
```
```suggestion
			$sValue = mb_substr($sValue, 0, $iMaxSize);
```

Molkobain replied:
+1

Hipska reviewed (changes requested):
Don't forget to also change the `substr` functions as you might get unexpected results otherwise..

accognet committed:
Add test

piRGoif commented on tests/php-unit-tests/unitary-tests/core/DBObjectTest.php:
```diff
@@ -897,4 +897,65 @@ private function CreatePersonInstance()
+	public function getCheckLongValueProvider()
+	{
+		return [
+			'Title with ééé'            => [
```
Seems like the use case in the original ticket, but you should describe it : something like 'Title with 200 é, failing before N°3448'

piRGoif replied:
Solved by Anne Catherine

piRGoif commented on tests/php-unit-tests/unitary-tests/core/DBObjectTest.php:
```diff
@@ -897,4 +897,65 @@ private function CreatePersonInstance()
+				'title',
+				'ééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééé',
+			],
+			'Title with smiley'         => [
```
Same, I would go with something like "title with 232 characters, ending with smileys (259 bytes)"

piRGoif replied:
Solved by Anne Catherine

piRGoif commented on tests/php-unit-tests/unitary-tests/core/DBObjectTest.php:
```diff
@@ -897,4 +897,65 @@ private function CreatePersonInstance()
+				'title',
+				'012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345😃😃😃😃😃😃😃😃😃',
+			],
+			'Title with 255 characters' => [
```
```suggestion
			'Title with 255 characters,  us-ascii only' => [
```

piRGoif commented on tests/php-unit-tests/unitary-tests/core/DBObjectTest.php:
```diff
@@ -897,4 +897,65 @@ private function CreatePersonInstance()
+				'title',
+				'ééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééé',
+			],
+		];
```
In the original ticket there was an issue also with loading an email in a 64KB blob fiels (I guess a TEXT ?)
It would be nice to test that too !

piRGoif replied:
Test rewritten with Anne Catherine yesterday, closing this conversation

piRGoif commented on tests/php-unit-tests/unitary-tests/core/DBObjectTest.php:
```diff
@@ -897,4 +897,65 @@ private function CreatePersonInstance()
+
+			$oPerson->Set('name', 'My last name');
+			$oPerson->Set('first_name', 'My first name');
+			$oPerson->Set('org_id', $this->getTestOrgId());*/
```
Remove code in comment

And use ItopDataTestCase::CreateUserRequest instead so that object can be rollbacked on tearDown !

piRGoif replied:
Solved by Anne Catherine

piRGoif reviewed (changes requested):
First, many thanks to propose a fix for this bug ! I don't like having it around, and the ticket not planned :/

I agree with Hipska suggestions...
I propose some modifications in the test.

We will discuss about your fix on next technical review, but I'm not sure why there are so much methods touched ? I would say, either we are touching only the code involved in the ticket use cases, or we replace any substr() and strlen() call by their mb_* equivalent ? There are lots remaining...

piRGoif commented:
As a reference, the mbstring functions list : https://www.php.net/manual/en/ref.mbstring.php

piRGoif commented:
> We will discuss about your fix on next technical review, but I'm not sure why there are so much methods touched ? I would say, either we are touching only the code involved in the ticket use cases, or we replace any substr() and strlen() call by their mb_* equivalent ? There are lots remaining...

One of them is solved in #529...

accognet committed:
Update core/dbobject.class.php

Co-authored-by: Thomas Casteleyn <thomas.casteleyn@super-visions.com>

accognet committed:
Update core/event.class.inc.php

Co-authored-by: Thomas Casteleyn <thomas.casteleyn@super-visions.com>

accognet committed:
Update core/event.class.inc.php

Co-authored-by: Thomas Casteleyn <thomas.casteleyn@super-visions.com>

accognet committed:
Update synchro/synchrodatasource.class.inc.php

Co-authored-by: Thomas Casteleyn <thomas.casteleyn@super-visions.com>

accognet committed:
Update tests/php-unit-tests/unitary-tests/core/DBObjectTest.php

Co-authored-by: Pierre Goiffon <pierre.goiffon@combodo.com>

accognet committed:
Update webservices/webservices.class.inc.php

Co-authored-by: Thomas Casteleyn <thomas.casteleyn@super-visions.com>

accognet committed:
Update synchro/synchrodatasource.class.inc.php

Co-authored-by: Thomas Casteleyn <thomas.casteleyn@super-visions.com>

accognet committed:
Update core/event.class.inc.php

Co-authored-by: Thomas Casteleyn <thomas.casteleyn@super-visions.com>

accognet committed:
N°3448 - Framework field size check not correctly implemented for multibytes languages/strings

piRGoif commented:
Technical review : Anne Catherine and Pierre will improve the test before functional review.

accognet committed:
Add tests

piRGoif commented:
Functional review : OK !
Note there is another PR on object deletion : #529

Before merge test review to be done by Pierre.

piRGoif commented:
Discussed with Anne Catherine : the test for the 64Ko should be simplified. 
Use case to test:
- a content < 64Ko
- a content > 64Ko that would cut normally before fix
- a content > 64Ko that would cut inside a multi bytes character before fix

Pass to the test only the content, and if value is bigger then also the expected cut value

Do the CheckToWrite and SetTrim (if CheckToWrite is returning false when content to big then just do the SetTrim in this case ?)

odain-cbd commented:
that's a real progress for our customers/our community! especially for those that speak arabic or asian languages....

accognet committed:
WIP

piRGoif committed:
Dataprovider rename cases

piRGoif commented:
Tests still to be reviewed. We began last monday, but Anne-Catherine and I will continue after our holidays.

piRGoif committed:
Tests review
- replace AttributeString & AttributeText tests
- kept EventIssue tests, add comment to explain why

piRGoif commented:
Tests reviewed ! Planned PR for functional review.

piRGoif committed:
DBObjectTest : fix wrongly named var

piRGoif commented:
We should rebase the PR, as its starting point is a commit made in... july !!

Molkobain commented:
Accepted during functional review for 3.1.2 if finished in time.
Still need to be reviewed technically.

piRGoif committed:
DBObjectTest::testCheckLongValueInAttribute : add a prefix so that emoji bytes won't be even

piRGoif committed:
DBObjectTest::testCheckLongValueInAttribute : rename (again...) var

accognet committed:
Remove duplicated test.
Rewrite one provider

piRGoif reviewed (approved):
Thanks for the rebase ! And also the duplicated test removal !

accognet commented on tests/php-unit-tests/unitary-tests/core/DBObjectTest.php:
```diff
@@ -1235,4 +1237,159 @@ public function testConstructorMemoryFootprint():void
+		if ($bIsValueToSetBelowAttrMaxSize) {
+			$this->assertEquals($sValueToSet, $sValueInObject);
+		} else {
+			$this->assertEquals($iAttrMaxSize, mb_strlen($sValueInObject));
```
```suggestion
			$this->assertEquals($iAttrMaxSize, mb_strlen($sValueInObject),'Should truncate at the same length than attribute max length');
```

accognet commented on tests/php-unit-tests/unitary-tests/core/DBObjectTest.php:
```diff
@@ -1235,4 +1237,159 @@ public function testConstructorMemoryFootprint():void
+		} else {
+			$this->assertEquals($iAttrMaxSize, mb_strlen($sValueInObject));
+			$sLastCharsOfValueInObject = mb_substr($sValueInObject, -30);
+			$this->assertStringContainsString(' -truncated', $sLastCharsOfValueInObject);
```
```suggestion
			$this->assertStringContainsString(' -truncated', $sLastCharsOfValueInObject, 'Should end with "truncated" comment');
```

accognet commented on tests/php-unit-tests/unitary-tests/core/DBObjectTest.php:
```diff
@@ -1235,4 +1237,159 @@ public function testConstructorMemoryFootprint():void
+		$oTicket->SetTrim($sAttrCode, $sValueToSet);
+		$sValueInObject = $oTicket->Get($sAttrCode);
+		if ($bIsValueToSetBelowAttrMaxSize) {
+			$this->assertEquals($sValueToSet, $sValueInObject);
```
```suggestion
			$this->assertEquals($sValueToSet, $sValueInObject,'Should not alter string that is already shorter than attribute max length');
```

accognet commented on tests/php-unit-tests/unitary-tests/core/DBObjectTest.php:
```diff
@@ -1235,4 +1237,159 @@ public function testConstructorMemoryFootprint():void
+		$bIsValueToSetBelowAttrMaxSize = ($iValueLength <= $iAttrMaxSize);
+		/** @noinspection PhpUnusedLocalVariableInspection */
+		[$bCheckStatus, $aCheckIssues, $bSecurityIssue] = $oTicket->CheckToWrite();
+		$this->assertEquals($bIsValueToSetBelowAttrMaxSize, $bCheckStatus, "CheckResult result:".var_export($aCheckIssues,true));
```
```suggestion
		$this->assertEquals($bIsValueToSetBelowAttrMaxSize, $bCheckStatus, "CheckResult result:".var_export($aCheckIssues, true));
```

accognet commented on tests/php-unit-tests/unitary-tests/core/DBObjectTest.php:
```diff
@@ -1235,4 +1237,159 @@ public function testConstructorMemoryFootprint():void
+
+		$oTicket->Set($sAttrCode, $sValueToSet);
+		$sValueInObject = $oTicket->Get($sAttrCode);
+		$this->assertSame($sValueToSet, $sValueInObject);
```
```suggestion
		$this->assertSame($sValueToSet, $sValueInObject, 'Set should not alter the value even if the value is too long');
```

accognet commented on tests/php-unit-tests/unitary-tests/core/DBObjectTest.php:
```diff
@@ -1235,4 +1237,159 @@ public function testConstructorMemoryFootprint():void
+	/**
+	 * Test check long field with non ascii characters
+	 *
+	 * @covers       DBObject::DBIncrement
```
```suggestion
	 * @covers       DBObject::Set
	 * @covers       DBObject::CheckToWrite
	 * @covers       DBObject::SetTrim
```

accognet commented on tests/php-unit-tests/unitary-tests/core/DBObjectTest.php:
```diff
@@ -1235,4 +1237,159 @@ public function testConstructorMemoryFootprint():void
+	/**
+	 * EventIssue has a OnInsert override that uses mb_strlen, so we need to test this specific case
+	 *
+	 * @covers       DBObject::DBIncrement
```
```suggestion
	 * @covers       EventIssue::OnInsert
```

accognet committed:
Apply suggestions from code review

piRGoif committed:
Move EventIssue test to a dedicated test class

Molkobain commented on tests/php-unit-tests/unitary-tests/core/DBObjectTest.php:
```diff
@@ -1235,4 +1237,97 @@ public function testConstructorMemoryFootprint():void
+	 */
+	public function testSetTrim($sName, $sResult){
+		$oOrganisation = MetaModel::NewObject(Organization::class);
+		$oOrganisation->SetTrim('name',$sName);
+		$this->assertEquals($sResult,$oOrganisation->Get('name'),'SetTrim must limit string to 255 characters');
```
```suggestion
		$oOrganisation->SetTrim('name', $sName);
		$this->assertEquals($sResult, $oOrganisation->Get('name'), 'SetTrim must limit string to 255 characters');
```

accognet committed:
code beautification

accognet committed:
code beautification

accognet committed:
code beautification
